configdir/interpolator.py

```diff
@@ -0,0 +1,92 @@
+import re
+from types import GeneratorType
+
+from .compat import Mapping, string_types, text_type
+
+
+class Interpolator(object):
+    """
+    Recursively interpolate variable placeholders in strings.
+
+    Strings can contain template place holders for other config keys. Placeholders
+    are enclosed with squiggly braces (``{{PLACEHOLDER}}``).
+
+    - Placeholders can refer to nested keys using dot (``.``)
+      notation: ``{{PARENT_KEY.NESTED_KEY}}``
+    - Placeholders can refer to nested list items using at (``@``)
+      notation: ``{{PARENT_KEY@1}}``
+    - Placeholders may only reference number or string values.
+    """
+
+    _key_regex = re.compile(r"{{(?P<key>[^\}]+)}}")
+
+    def __init__(self, values):
+        """
+        Args:
+            values (dict): Values that placeholders will be replaced with.
+        """
+        self.values = values
+
+    def _key_segments(self, full_key):
+        """
+        Yields:
+             tuple[str, str|int, list[str]]: Key segment prefix ("." or "@"), key
+                segment value, and list of current key path segments.
+        """
```

Comment:
Not sure if an example is needed or what, but this docstring does not really help me understand what this is for. Specifically, I'm not completely following the `path` part.

Reply:
`some.key@1` has path segments `[".", "some"]`, then `[".", "some", ".", "key"]` then `[".", "some", ".", "key", "@", 1]`